addons/mail/controllers/main.py

```diff
@@ -52,7 +52,9 @@ def _redirect_to_record(cls, model, res_id, access_token=None, **kwargs):
         # access_token and kwargs are used in the portal controller override for the Send by email or Share Link
         # to give access to the record to a recipient that has normally no access.
         uid = request.session.uid
-
+        user = request.env['res.users'].sudo().browse(uid)
+        allowed_company_ids = ','.join([str(cid) for cid in user.company_ids.ids])
```

Comment:
add a comment to explain why we need to extend the "allowed_company_ids" to allow a redirection to any record that the user can access, regardless of currently visible records based on the "currently allowed companies".
(the var could be renamed `cids` for brevity here)